### PR TITLE
Adds setting to disable cursor movement while typing into Find Widget

### LINKED ISSUE #60977: Disable Auto Searching while typing
Is there a way to disable automatic searching? I work in a 40k+ line document daily and searching through it constantly while typing (going up, down and all around) is a mess. I want Visual Studio Code to start searching only when I press enter/return on the keyboard and quit searching while I am typing what I want to search for! Where is this feature? Thanks in advance.

### LINKED ISSUE #70306: Editor jumps to random line when clicking into it
Issue Type: <b>Bug</b>

1. Open a repo in VS Code.
2. "Find in Files" for some string.
3. Click on a search result so it opens in a new editor window.
4. Click on a line of code in this new editor window.

Expected results:
The cursor is where I clicked. The window never scrolls anywhere.

Actual results:
Often enough to be annoying, but not reliably enough to describe it, the editor scrolls to a random location in the file. 

I'm sorry that I can't give you more details, but this bug is EXTREMELY annoying. If I could ask for a single thing in VS Code, it would be that you drop everything else you're working on and fix this bug. VS Code is a fantastic product, and I'm sure you have lots of ideas how to make it even better. But for my daily work, the one thing that I need is this bugfix.

VS Code version: Code 1.32.0 (507312a3e3b34b084b467dfd983263bc7c9d87e0, 2019-03-06T18:52:50.880Z)
OS version: Darwin x64 18.5.0

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Xeon(R) W-2191B CPU @ 2.30GHz (36 x 2300)|
|GPU Status|2d_canvas: enabled<br>checker_imaging: disabled_off<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>native_gpu_memory_buffers: enabled<br>rasterization: enabled<br>surface_synchronization: enabled_on<br>video_decode: enabled<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|1, 2, 2|
|Memory (System)|32.00GB (2.80GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|

</details><details><summary>Extensions (5)</summary>

Extension|Author (truncated)|Version
---|---|---
gitlens|eam|9.5.1
python|ms-|2019.2.5558
cpptools|ms-|0.21.0
vsliveshare|ms-|0.3.1284
vscode-capnp|xmo|1.0.0

</details>

### PR DESCRIPTION
**Pull request by ultraGentle**
This PR fixes #60977 and fixes #70306, both related to disorienting or disruptive cursor movement related to the Find Widget.  

Description: 
- Adds a boolean setting to turn cursor movement on/off while typing into the Find Widget.  
- Default is true (on), which results in the current behavior. 
- False (off) results in highlighting-only on type, preventing the undesired/unexpected cursor movement described in 60977 and 70306.  Jumping to the next match happens on `Enter` or on `click` the prev/next arrows in the widget.  

Discussion:
- The VSC team suggestion was to name the setting "searchOnType," however there is already a setting with that name.  I chose "moveOnType."  
- Personally, I would reverse the defaults and simply note the option to switch back in "What's New," but I can understand not wanting to confuse existing users.  
- The VSC team suggested that the setting have three options: `off | highlight | move`.   However, when I tested `off`, the behavior was just confusing, because it seemed like the Find Widget wasn't doing anything at all.  Also, from searching related issues on GitHub, I haven't seen anyone asking for the "off" behavior.  I thought it better not to overengineer a confusing 3rd setting that no one asked for.  

Further improvements:
- 70306 is only fixed when the new setting is activated (i.e. false).  It could be addressed separately, but there is already a workaround discussed in that thread, and for anyone who doesn't want to do that, there is the option provided by this PR.  
- For this setting to be more easily discovered, I would suggest adding the words "jump" and "scroll" to the description, because those were what folks used when describing the issue (so that's probably what others would search for).  

Any feedback is welcome, and I am happy for you to revise as you see fit.  
Thanks for considering!

### CODE REVIEW

File: src/vs/editor/contrib/find/findModel.ts [hunk 1]
```diff
@@ -3,26 +3,26 @@
...
-import { EndOfLinePreference, FindMatch, ITextModel } from 'vs/editor/common/model';
-import { SearchParams } from 'vs/editor/common/model/textModelSearch';
-import { FindDecorations } from 'vs/editor/contrib/find/findDecorations';
-import { FindReplaceState, FindReplaceStateChangedEvent } from 'vs/editor/contrib/find/findState';
-import { ReplaceAllCommand } from 'vs/editor/contrib/find/replaceAllCommand';
-import { ReplacePattern, parseReplaceString } from 'vs/editor/contrib/find/replacePattern';
-import { RawContextKey } from 'vs/platform/contextkey/common/contextkey';
-import { IKeybindings } from 'vs/platform/keybinding/common/keybindingsRegistry';
-import { EditorOption } from 'vs/editor/common/config/editorOptions';
+import {RunOnceScheduler, TimeoutTimer} from 'vs/base/common/async';
```
Comment (rebornix) [inline src/vs/editor/contrib/find/findModel.ts:6]: It seems the file is formatted by some formatters accidentally

**Review by rebornix — CHANGES_REQUESTED** [review]
It seems the file is formatted by some formatters accidentally

**Comment by ultraGentle** [comment]
Sorry about that.  

Is there a command I can run to apply your in-house formatting automatically/retroactively?

**Comment by ultraGentle** [comment]
@rebornix I believe I fixed the formatting issue.  Any next steps you need from me?

**Comment by briankendall** [comment]
I just tried checking out and building this PR, and it does not fix #70306 for me. Even with "search on type" disabled, when I click into another editor

**Comment by ultraGentle** [comment]
@briankendall Can you verify that you un-selected `move on type`, and not 'search on type`?

There is an unrelated, pre-existing setting `search on type` that has no effect.  

With `move on type` `off`, (i.e., implementing this PR) I can no longer reproduce #70306.

**Comment by briankendall** [comment]
@ultraGentle You're right, I changed the wrong setting! It is in fact working for me.

**Comment by ultraGentle** [comment]
@rebornix I just made another commit: 

- Debounces typing into the Find Widget  

- This way, even users that do *not* disable `moveOnType` will avoid the editor auto-searching until they finish/pause typing -- which seems universally desirable.   

- If you just want the `moveOnType` setting without debounce, please merge the previous commit instead.  

Any comments or feedback appreciated!

**Comment by adamreisnz** [comment]
Any update on this getting merged?

**Comment by skplunkerin** [comment]
Same, I'd love to have this

**Comment by ultraGentle** [comment]
I haven't heard from @rebornix that any changes are necessary, and as far as I can tell, all testers have had positive experiences with no issues.  Hoping this will make it into next month's update!

**Comment by adamreisnz** [comment]
Ok that would be great, looking forward to it.

**Comment by ultraGentle** [comment]
@adamreisnz , I think I implied that this *will* make it into the next release.  If so, sorry -- that was just me hoping along with everyone else on this thread!  I unfortunately don't have any insider info to share.

**Comment by dzoba** [comment]
Excited and hopeful to have this merged.

**Comment by ultraGentle** [comment]
I'm not sure what the deal is with radio silence from the VSCode team regarding this issue.  I've already pinged the reviewer, with no response, and I don't want to be a squeaky wheel, so it's not clear to me what the way forward is.  If I knew there were a problem, I'd fix it, but without any feedback I'm flummoxed.  

It's very strange to me, since this is a small PR and it fixes a super annoying bug, but I know folks are busy, so maybe there just hasn't been time for them to re-review and merge?  Or maybe there's something else holding it up?  

The opacity here is a disincentive to contribute to other issues, since it feels futile to work on something if it's not going to end up in the final product.  Not complaining, promise, just confused!

**Comment by adamreisnz** [comment]
@rebornix would you mind reviewing this PR? It would really save us a lot of headaches with the cursor jumping to unexpected locations in the file.

**Comment by rebornix** [comment]
Thanks @ultraGentle for your contribution! I tweaked the code a bit and merged into master, should be available in Insiders later this week.

**Comment by ultraGentle** [comment]
@rebornix My pleasure!

**Comment by TheInvoker** [comment]
Is this implemented? How can I enable this setting?

**Comment by ultraGentle** [comment]
@TheInvoker Should be as of today or tomorrow.  It's been in insiders edition for a month, so just waiting for 1.49 to be released.  Setting is find > "moveCursorOnType" if I remember correctly.

**Comment by ultraGentle** [comment]
We're finally there!  

1.49.0 stable was released today with the setting 
`find` > `moveCursorOnType` > set to `false` 
which activates the desired behavior

Edit: Note that if you previously had `global find clipboard` set to false to avoid unexpected cursor jumping/highlighting when switching panes, you can now re-enable that setting, as `moveCursorOnType` > `false` fixes that bug.

**Pull request merged**